ocaml/typing/typecore.ml

```diff
@@ -6546,15 +6578,66 @@ and type_expect_
       end;
       let exp_extra = (Texp_stack, loc, []) :: exp.exp_extra in
       {exp with exp_extra}
-  | Pexp_overwrite (_x, _e) ->
+  | Pexp_overwrite (exp1, exp2) ->
       if not (Language_extension.is_enabled Overwriting) then
-        raise (Typetexp.Error (loc, env,
-                               Unsupported_extension Overwriting))
-      else Location.todo_overwrite_not_implemented loc
+        raise (Typetexp.Error (loc, env, Unsupported_extension Overwriting));
+      if not (can_be_overwritten exp2.pexp_desc) then
+        raise (Syntaxerr.(Error(Expecting(exp2.pexp_loc, "tuple, constructor or record"))));
```

Comment:
This is a tough call. I agree that it would be cleaner to make this a type error. But a type error can only report an error location in a typed term, which would require us to check the whole expression before we see that the given term can not appear as an argument to overwrite. I think this would give a worse user-experience than the current approach: For example, a confused user might write `overwrite_ e with stack_ (a, _)`. If we type-check the inner term first, the error message will be `unexpected wildcard "_"`, which is not very helpful.

Reply:
I agree both with your prioritising UX here, and that it's better to report the "expected a tuple, etc" error. But I don't see why we would need to type-check the expression before reporting a type error. You'd probably have to add a new constructor to the `error` type at the top of the file, but you can put whatever you like in the new constructor. Is there something I'm missing here?